File: P4/NapakalakiRuby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module NapakalakiRuby
  class SpecificBadConsequence < BadConsequence
    attr_reader :text, :levels, :specificVisibleTreasures, :specificHiddenTreasures
    
    def initialize(aText, someLevels, someSpecificVisibleTreasures, someSpecificHiddenTreasures)
      @text=aText
      @levels=someLevels
      @specificVisibleTreasures=someSpecificVisibleTreasures
      @specificHiddenTreasures=someSpecificHiddenTreasures 
    end
    
    def isEmpty
      empty = false      
      if @specificVisibleTreasures.empty? && @specificHiddenTreasures.empty? then
        empty = true
      end
      return empty
    end

    def substractVisibleTreasure(t)
      @specificVisibleTreasures.delete(t.type)
    end
       #Modificado
    def substractHiddenTreasure(t)
      @specificHiddenTreasures.delete(t.type)
    end
    
    def adjustToFitTreasureLists(v, h)
      treasuresVisible=Array.new
      treasuresHidden=Array.new
      bad=SpecificBadConsequence.new("", 1, 0, 0)
     
      if !@specificHiddenTreasures.empty? || !@specificVisibleTreasures.empty?
        
        if !v.empty?
            v.each{|tV|
              if !treasuresVisible.include?(tV.type)
                  treasuresVisible<<tV.type
              end
           }
        end
        if !h.empty?
          h.each{|tH|
              if !treasuresHidden.include?(tH.type)
                  treasuresHidden<<tH.type
              end
           }
        end
        bad=SpecificBadConsequence.new( @text, 0, treasuresVisible,treasuresHidden); 
      end
        return bad
    end
    
    def to_s
        "Tesoros visibles especificos = #{@specificHiddenTreasures}"  + 
        "Tesoros ocultos especificos = #{@specificVisibleTreasures}" 
    end

  end
end
 

File: P4/Napakalaki/src/napakalaki/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;
import java.util.ArrayList;
import java.util.Collections;
/**
 *
 * @author pako
 */
public class SpecificBadConsequence extends BadConsequence{
    ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList(); 
    ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    
    public SpecificBadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden)
    {
        super(text, levels, tVisible, tHidden);
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures(){
        return specificHiddenTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificVisibleTreasures(){
        return specificVisibleTreasures;
    }
    
    @Override
    public boolean isEmpty()
    {
      return this.specificVisibleTreasures.isEmpty() && this.specificHiddenTreasures.isEmpty();
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t)
    {
      specificVisibleTreasures.remove(t.getType());
    }

    @Override
    public void substractHiddenTreasure(Treasure t)
    {
      specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v ,ArrayList<Treasure> h ){
        ArrayList<TreasureKind> treasuresVisible = new ArrayList();
        ArrayList<TreasureKind> treasuresHidden = new ArrayList();
        SpecificBadConsequence bad=new SpecificBadConsequence("",1,new ArrayList(),new ArrayList());   
        
            if(!this.specificHiddenTreasures.isEmpty()  || !this.specificVisibleTreasures.isEmpty()  )
            {
                if ( !v.isEmpty())
                {
                    for (Treasure t: v)
                    {
                        //Si no contiene el tipo de tesoro t lo agregamos al nuevo ArrayList de visibles
                        if (!treasuresVisible.contains(t.getType()))   //.contains() --> metodo propio de ArrayList
                            treasuresVisible.add(t.getType());           //Agregamos el el nuevo tipo de tesoro al nuevo ArrayList
                    
                     }
                }
                if ( !h.isEmpty())
                {

                    //Recorremos los tesoros ocultos
                    for (Treasure t: h)
                    {
                        //Si no contiene el tipo de tesoro t lo agregamos al nuevo ArrayList de ocultos
                        if (!treasuresHidden.contains(t.getType()))    //.contains() --> metodo propio de ArrayList
                            treasuresHidden.add(t.getType());
                        
                    }

                }
                bad = new SpecificBadConsequence(text, 0, treasuresVisible, treasuresHidden);
            }
            return bad;
    }
    
    @Override
    public String toString(){
        return "Tesoros visibles especificos =" + this.specificVisibleTreasures +
                "Tesoros ocultos especificos =" + this.specificHiddenTreasures;
    }
}
